### DISCUSSION TITLE
Security concerns with using react-router's navigation function with values from localStorage

### DISCUSSION THREAD

**Discussion by artemhp** [remix-run/remix#6240] [Q&A]
Hi everyone,

I am working on a React application that uses react-router for client-side routing. I am currently using the `useNavigation` hook to navigate between different pages in my application, and I am retrieving a value from `localStorage` to pass to the navigation function.

However, I recently ran a security scan on my application using `Fortify`, and it flagged the use of react-router's navigation function as a potential security issue. Fortify suggested that it is not secure to use navigation functions with values retrieved from `localStorage`.

I am wondering if anyone has any insights or best practices for using react-router's navigation function securely with values retrieved from `localStorage`. Are there any built-in features or best practices that can help me prevent security vulnerabilities when using react-router's navigation function?

Any advice or feedback would be greatly appreciated. Thanks!

**Comment by sergiodxa** [comment]
That makes sense, the value in localStorage can be modified by the user or even any other script loading in your app, including compromised third-party.

So instead of using directly the value you probably want to ensure it's a valid value, like it's one of your routes and it's not fallback to a safe value.

**Comment by artemhp** [comment]
@sergiodxa 
Thanks for the answer! 
Just to clarify, does react-router's `navigation` function perform any validation or sanitization to prevent potential XSS attacks, unknown routes, or external URLs from being passed as parameters? Or is it up to the developer to ensure that any values passed to the navigation function are properly validated and sanitized?

Thanks again for your help!

**Reply by sergiodxa** [reply]
It doesn't, but also you can only use it to navigate to another client-side route, not to another domain AFAIK.
